lamindb_setup/dev/upath.py

```diff
@@ -402,24 +383,57 @@ def inner(dir_path: Union[Path, UPath], prefix: str = "", level: int = -1):
 """
 
 
-def create_path(pathlike: Union[str, Path, UPath]) -> UPath:
+def convert_pathlike(pathlike: Union[str, Path, UPath]) -> UPath:
     """Convert pathlike to Path or UPath inheriting options from root."""
     if isinstance(pathlike, (str, UPath)):
         path = UPath(pathlike)
     elif isinstance(pathlike, Path):
         path = UPath(str(pathlike))  # UPath applied on Path gives Path back
     else:
         raise ValueError("pathlike should be of type Union[str, Path, UPath]")
-
     # remove trailing slash
     if path._parts[-1] == "":
         path._parts = path._parts[:-1]
+    return path
 
-    if isinstance(path, S3Path):
-        path = UPath(path, cache_regions=True)
-        if AWS_CREDENTIALS_PRESENT is None:
-            set_aws_credentials_present(path)
-        if not AWS_CREDENTIALS_PRESENT:
-            path = UPath(path, anon=True)
 
-    return path
+def create_path(path: UPath) -> UPath:
+    global AWS_CREDENTIALS_PRESENT
+
+    path = convert_pathlike(path)
+    if not isinstance(path, S3Path):
+        return path
+    path = UPath(path, cache_regions=True)
+    # the below is problematic, because it will assume that all subsequent
+    # requests will then be treated as anon, and we'll never try something else
+    # if isinstance(path, S3Path) and not upath.AWS_CREDENTIALS_PRESENT:
+    #     # if we already know we don't have AWS credentials
+    #     # let's make an anon request from the get-go
+    #     return UPath(path, anon=True)
+    try:
+        path.fs.call_s3("head_bucket", Bucket=path._url.netloc)
+        AWS_CREDENTIALS_PRESENT = True  # type: ignore
```

Comment:
Not sure this makes sense. The point of AWS_CREDENTIALS_PRESENT is to avoid path.fs.call_s3 on every call of create_path. So if it is set, the anon check should not be done. This is how the original create_path function does this. Now create_path calls path.fs.call_s3 every time.

Reply:
I think the correct logic is.
Make a global AWS_MANAGED_CREDENTIALS=None.
if neither AWS_MANAGED_CREDENTIALS nor AWS_CREDENTIALS_PRESENT  set, check with path.fs.call_s3. 
If fails, check for AWS_MANAGED_CREDENTIALS. If success, set AWS_MANAGED_CREDENTIALS=True and do the reuired logic. If fails, check for AWS_CREDENTIALS_PRESENT  and do the required logic with anon.
if both variables set, just return the path.
This looks the correct way to me. But maybe i just don't understand what needs to be achieved here.

Reply:
Anyways the current variant doesn't look correct to me.

Reply:
The point here is that this check should not be done for every UPath conversion with create_path in lamindb.

Reply:
Please read the comment! I'll point you to it.

Reply:
It's the comment in the code that ends 3 line above your GitHub comment :D

I tried to answer in this comment what you're pointing out here:
<img width="685" alt="image" src="https://github.com/laminlabs/lamindb-setup/assets/16916678/e6d8c618-cf5e-40a1-996e-a67592761d83">

Reply:
Let me know if the reasoning makes sense!

Reply:
I think the repeated requests are OK because the canonical usage will center around the default storage.

I'm sure there is optimization potential here, but I think the main challenge will be to enable to work across different sets of credentials.

Reply:
Not sure. This makes the head_bucket check every time create_path is called, and it is called a lot in lamindb.

Reply:
But the whole thing is pretty hard to understand already as the PR has too many things and some them can affect the credentials logic also.

Reply:
Do you see the basic problem that we'll never hit the `access-aws` check in case there are no credentials? That means: people without AWS credentials in their environment would never get credentials for our hosted storage?

Reply:
That's why we can't do the naive AWS_CREDENTIALS_PRESENT check we did before in the same way anymore.

Reply:
Yes, i see it. The problem is current logic though is that it won't update the root https://github.com/laminlabs/lamindb-setup/blob/a03dfe2f2970752e4392c6302631eb2b400a6f63/lamindb_setup/dev/_settings_storage.py#L173 in this case.

So what happens as i understand it when there are no credentials.
the root is set with anon=True and is not updated.

also setting the environment variables seems useless to me if `s3fs.S3FileSystem.cachable = False` is not set due to s3fs caching behaviour described in my comments.

Reply:
Agreed about setting the environment variables being useless. But I also think it doesn't hurt right now and it's nice to test things (it's a nice overwriting behavior upon calling `access_aws`).